pkg/sensors/accelerator/accelerator.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2024.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package accelerator
+
+//nolint:gci // The supported device imports are kept separate.
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sustainable-computing-io/kepler/pkg/sensors/accelerator/device"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	// Add supported devices.
+	_ "github.com/sustainable-computing-io/kepler/pkg/sensors/accelerator/device/sources"
+)
+
+const (
+	DUMMY = iota
+	GPU
+	// Add other accelerator types here
+)
+
+var (
+	globalRegistry *AcceleratorRegistry
+	once           sync.Once
+)
+
+// Accelerator represents an implementation of... equivalent Accelerator device.
+type Accelerator interface {
+	// Device returns an underlying accelerator device implementation...
+	Device() device.DeviceInterface
+	// DeviceType ...
+	DeviceType() device.DeviceType
+	// IsRunning ...
+	IsRunning() bool
+	// StopAccelerator ...
+	Stop()
+	AccType() AcceleratorType
+}
+
+type AcceleratorType int
+
+type AcceleratorRegistry struct {
+	m        sync.Mutex
+	Registry map[device.DeviceType]Accelerator
+}
+type accelerator struct {
+	dev           device.DeviceInterface // Device Accelerator Interface
+	accType       AcceleratorType
+	running       bool
+	installedTime metav1.Time
+}
+
+func (a AcceleratorType) String() string {
+	return [...]string{"DUMMY", "GPU"}[a]
+}
+
+// Registry gets the default device AcceleratorRegistry instance
+func Registry() *AcceleratorRegistry {
+	once.Do(func() {
+		globalRegistry = &AcceleratorRegistry{
+			m:        sync.Mutex{},
+			Registry: map[device.DeviceType]Accelerator{},
+		}
+	})
+	return globalRegistry
+}
+
+// SetRegistry replaces the global registry instance
+// NOTE: All plugins will need to be manually registered
+// after this function is called.
+func SetRegistry(registry *AcceleratorRegistry) {
+	globalRegistry = registry
+}
+
+// Devices returns a map of supported accelerators.
+func (r *AcceleratorRegistry) Accelerators() (map[device.DeviceType]Accelerator, error) {
+	if len(r.Registry) == 0 {
```

Comment:
I see that we are using `mutex` to protect `r.Registry` but there places where `Registry` map is accessed and exposed to the world outside. This isn't thread safe. We have to make a decision if the `Registry` needs to be threadsafe and prevent access to internal members though mutex (preferably RWMutex) or use `sync.Map`